driver-core/src/test/functional/com/mongodb/client/model/expressions/TypeExpressionsFunctionalTest.java

```diff
@@ -0,0 +1,244 @@
+        assertExpression(1, of(1).isNumberOr(of(99)), "{'$cond': [{'$isNumber': [1]}, 1, 99]}");
+        // other numeric values:
+        assertExpression(1L, of(1L).isNumberOr(of(99)));
+        assertExpression(1.0, of(1.0).isNumberOr(of(99)));
+        assertExpression(Decimal128.parse("1"), of(Decimal128.parse("1")).isNumberOr(of(99)));
+        // non-numeric:
+        assertExpression(99, ofIntegerArray(1).isNumberOr(of(99)));
+        assertExpression(99, ofNull().isNumberOr(of(99)));
+    }
+
+    @Test
+    public void isStringOrTest() {
+        assertExpression(
+                "abc",
+                of("abc").isStringOr(of("or")),
+                "{'$cond': [{'$eq': [{'$type': 'abc'}, 'string']}, 'abc', 'or']}");
+        // non-string:
+        assertExpression("or", ofIntegerArray(1).isStringOr(of("or")));
+        assertExpression("or", ofNull().isStringOr(of("or")));
+    }
+
+    @Test
+    public void isDateOrTest() {
+        Instant date = Instant.parse("2007-12-03T10:15:30.005Z");
+        assertExpression(
+                date,
+                of(date).isDateOr(of(date.plusMillis(10))),
+                "{'$cond': [{'$in': [{'$type': {'$date': '2007-12-03T10:15:30.005Z'}}, ['date']]}, "
+                        + "{'$date': '2007-12-03T10:15:30.005Z'}, {'$date': '2007-12-03T10:15:30.015Z'}]}");
+        // non-date:
+        assertExpression(date, ofIntegerArray(1).isDateOr(of(date)));
+        assertExpression(date, ofNull().isDateOr(of(date)));
+    }
+
+    @Test
+    public void isArrayOrTest() {
+        // https://www.mongodb.com/docs/manual/reference/operator/aggregation/isArray/ (36 |47)
+        assertExpression(
+                Arrays.asList(1, 2),
+                ofIntegerArray(1, 2).isArrayOr(ofIntegerArray(99)),
+                "{'$cond': [{'$isArray': [[1, 2]]}, [1, 2], [99]]}");
+        // non-array:
+        assertExpression(Arrays.asList(1, 2), of(true).isArrayOr(ofIntegerArray(1, 2)));
+        assertExpression(Arrays.asList(1, 2), ofNull().isArrayOr(ofIntegerArray(1, 2)));
+    }
+
+    @Test
+    public void isDocumentOrTest() {
+        BsonDocument doc = BsonDocument.parse("{a: 1}");
+        assertExpression(doc,
+                of(doc).isDocumentOr(of(BsonDocument.parse("{b: 2}"))),
+                "{'$cond': [{'$eq': [{'$type': {'$literal': {'a': 1}}}, 'object']}, "
+                        + "{'$literal': {'a': 1}}, {'$literal': {'b': 2}}]}");
+        // non-document:
+        assertExpression(doc, ofIntegerArray(1).isDocumentOr(of(doc)));
+        assertExpression(doc, ofNull().isDocumentOr(of(doc)));
+    }
+
+    // conversions
+    // https://www.mongodb.com/docs/manual/reference/operator/aggregation/convert/ (38 |40)
+    // Convert is not implemented: too dynamic, conversions should be explicit.
+
+    /*
+    One might expect to see all conversions in $convert represented in this
+    API, but we expose only the useful ones.
+
+    Useful conversions:
```

Comment:
I find comments like these in test files to be unhelpful.  They should be moved to the Jira issue or the commit message to explain the rationale for decisions that we made.  For anyone that cares, this is about the last place anyone would thing to look.

Reply:
Removed the ones I am aware of.